File: programs/sss-token/src/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct StablecoinInitialized {
    pub mint: Pubkey,
    pub authority: Pubkey,
    pub name: String,
    pub symbol: String,
    pub decimals: u8,
    pub supply_cap: u64,
    pub enable_permanent_delegate: bool,
    pub enable_transfer_hook: bool,
    pub timestamp: i64,
}

#[event]
pub struct TokensMinted {
    pub mint: Pubkey,
    pub recipient: Pubkey,
    pub amount: u64,
    pub new_supply: u64,
    pub minter: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TokensBurned {
    pub mint: Pubkey,
    pub amount: u64,
    pub new_supply: u64,
    pub burner: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AccountFrozenEvent {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AccountThawedEvent {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct StablecoinPaused {
    pub mint: Pubkey,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct StablecoinUnpaused {
    pub mint: Pubkey,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct BlacklistAdded {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub reason: String,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct BlacklistRemoved {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TokensSeized {
    pub mint: Pubkey,
    pub from: Pubkey,
    pub to: Pubkey,
    pub amount: u64,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct RoleUpdated {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub role: String,
    pub granted: bool,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct MinterUpdated {
    pub mint: Pubkey,
    pub minter: Pubkey,
    pub quota: u64,
    pub active: bool,
    pub authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AuthorityNominated {
    pub mint: Pubkey,
    pub current_authority: Pubkey,
    pub pending_authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AuthorityAccepted {
    pub mint: Pubkey,
    pub old_authority: Pubkey,
    pub new_authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct SupplyCapUpdated {
    pub mint: Pubkey,
    pub old_cap: u64,
    pub new_cap: u64,
    pub authority: Pubkey,
    pub timestamp: i64,
}

File: sdk/src/events.ts
import { PublicKey } from "@solana/web3.js";
import { BorshCoder, EventParser, Program } from "@coral-xyz/anchor";
import BN from "bn.js";

// Event type definitions matching the IDL
export interface StablecoinInitializedEvent {
  name: "stablecoinInitialized";
  data: {
    config: PublicKey;
    mint: PublicKey;
    masterAuthority: PublicKey;
    name: string;
    symbol: string;
    preset: number;
    timestamp: BN;
  };
}

export interface TokensMintedEvent {
  name: "tokensMinted";
  data: {
    config: PublicKey;
    minter: PublicKey;
    recipient: PublicKey;
    amount: BN;
    totalMinted: BN;
    timestamp: BN;
  };
}

export interface TokensBurnedEvent {
  name: "tokensBurned";
  data: {
    config: PublicKey;
    burner: PublicKey;
    from: PublicKey;
    amount: BN;
    totalBurned: BN;
    timestamp: BN;
  };
}

export interface AccountFrozenEvent {
  name: "accountFrozen";
  data: {
    config: PublicKey;
    authority: PublicKey;
    targetAccount: PublicKey;
    timestamp: BN;
  };
}

export interface AccountThawedEvent {
  name: "accountThawed";
  data: {
    config: PublicKey;
    authority: PublicKey;
    targetAccount: PublicKey;
    timestamp: BN;
  };
}

export interface ProgramPausedEvent {
  name: "programPaused";
  data: {
    config: PublicKey;
    pauser: PublicKey;
    timestamp: BN;
  };
}

export interface ProgramUnpausedEvent {
  name: "programUnpaused";
  data: {
    config: PublicKey;
    pauser: PublicKey;
    timestamp: BN;
  };
}

export interface RoleUpdatedEvent {
  name: "roleUpdated";
  data: {
    config: PublicKey;
    role: string;
    oldHolder: PublicKey;
    newHolder: PublicKey;
    updatedBy: PublicKey;
    timestamp: BN;
  };
}

export interface MinterUpdatedEvent {
  name: "minterUpdated";
  data: {
    config: PublicKey;
    minter: PublicKey;
    isActive: boolean;
    mintQuota: BN;
    updatedBy: PublicKey;
    timestamp: BN;
  };
}

export interface AuthorityTransferredEvent {
  name: "authorityTransferred";
  data: {
    config: PublicKey;
    oldAuthority: PublicKey;
    newAuthority: PublicKey;
    timestamp: BN;
  };
}

export interface AuthorityNominatedEvent {
  name: "authorityNominated";
  data: {
    config: PublicKey;
    oldAuthority: PublicKey;
    nominatedAuthority: PublicKey;
    timestamp: BN;
  };
}

export interface BlacklistAddedEvent {
  name: "blacklistAdded";
  data: {
    config: PublicKey;
    blockedAddress: PublicKey;
    reason: string;
    blacklistedBy: PublicKey;
    timestamp: BN;
  };
}

export interface BlacklistRemovedEvent {
  name: "blacklistRemoved";
  data: {
    config: PublicKey;
    unblockedAddress: PublicKey;
    removedBy: PublicKey;
    timestamp: BN;
  };
}

export interface AllowlistAddedEvent {
  name: "allowlistAdded";
  data: {
    config: PublicKey;
    address: PublicKey;
    addedBy: PublicKey;
    reason: string;
    timestamp: BN;
  };
}

export interface AllowlistRemovedEvent {
  name: "allowlistRemoved";
  data: {
    config: PublicKey;
    address: PublicKey;
    removedBy: PublicKey;
    timestamp: BN;
  };
}

export interface TokensSeizedEvent {
  name: "tokensSeized";
  data: {
    config: PublicKey;
    from: PublicKey;
    amount: BN;
    seizedBy: PublicKey;
    timestamp: BN;
  };
}

export interface AuditLogRecordedEvent {
  name: "auditLogRecorded";
  data: {
    config: PublicKey;
    index: BN;
    action: number;
    actor: PublicKey;
    timestamp: BN;
  };
}

export interface SupplyCapUpdatedEvent {
  name: "supplyCapUpdated";
  data: {
    config: PublicKey;
    oldCap: BN;
    newCap: BN;
    timestamp: BN;
  };
}

export interface MetadataUpdatedEvent {
  name: "metadataUpdated";
  data: {
    config: PublicKey;
    timestamp: BN;
  };
}

export type SSSEvent =
  | StablecoinInitializedEvent
  | TokensMintedEvent
  | TokensBurnedEvent
  | AccountFrozenEvent
  | AccountThawedEvent
  | ProgramPausedEvent
  | ProgramUnpausedEvent
  | RoleUpdatedEvent
  | MinterUpdatedEvent
  | AuthorityTransferredEvent
  | AuthorityNominatedEvent
  | BlacklistAddedEvent
  | BlacklistRemovedEvent
  | AllowlistAddedEvent
  | AllowlistRemovedEvent
  | TokensSeizedEvent
  | AuditLogRecordedEvent
  | SupplyCapUpdatedEvent
  | MetadataUpdatedEvent;

export function createEventParser(program: Program): EventParser {
  return new EventParser(program.programId, new BorshCoder(program.idl));
}

export function parseTransactionEvents(
  program: Program,
  logs: string[]
): SSSEvent[] {
  const parser = createEventParser(program);
  const events: SSSEvent[] = [];
  for (const event of parser.parseLogs(logs)) {
    events.push(event as unknown as SSSEvent);
  }
  return events;
}
